axiom/_schema.py

```diff
@@ -3,28 +3,33 @@
 CREATE_OBJECT = 'INSERT INTO *DATABASE*.axiom_objects (type_id) VALUES (?)'
 CREATE_TYPE = 'INSERT INTO *DATABASE*.axiom_types (typename, module, version) VALUES (?, ?, ?)'
 
+GET_TABLE_INFO = 'PRAGMA *DATABASE*.table_info(?)'
 
-BASE_SCHEMA = ["""
+
+# AUTOINCREMENT so that oids are never reused
+CREATE_OBJECTS = """
 CREATE TABLE *DATABASE*.axiom_objects (
+    oid INTEGER PRIMARY KEY AUTOINCREMENT,
```

Comment:
https://sqlite.org/autoinc.html states "The AUTOINCREMENT keyword imposes extra CPU, memory, disk space, and disk I/O overhead and should be avoided if not strictly needed. It is usually not needed."  Further reading of that document suggests that we don't need AUTOINCREMENT here, and that simply declaring the column as being INTEGER PRIMARY KEY will cause sqlite to do the right thing.

Reply:
Would it make sense to call the column "storeID", since that is how the rest of Axiom's codebase refers to it?

Reply:
We need AUTOINCREMENT here to avoid reusing storeIDs belonging to deleted objects. (Currently these rows are never deleted, so their reuse is prevented that way, but I think being able to reclaim the disk space and using AUTOINCREMENT is preferrable to the current scenario)

I'll check to see if you can still refer to the column as `oid` if it is named `storeID`.

Reply:
Actually on second thought, it'll be more confusing to have the column named differently to how it is referenced everywhere, and I don't want to have to check which name to use all the time.

Reply:
According to the sqlite docs, "If a table contains a column of type INTEGER PRIMARY KEY, then that column becomes an alias for the ROWID. You can then access the ROWID using any of four different names, the original three names described above or the name given to the INTEGER PRIMARY KEY column."

Based on that quote, I would say "storeID" should just work.

Reply:
Continuing the AUTOINCREMENT discussion, why are we concerned about the reuse of rowid values within a single table?  I understand the original problem, which is that tables without an  explicitly declared INTEGER PRIMARY KEY will have their rowids re-written when VACUUMed; but with the column being explicitly declared, existing rows won't have their rowids changed.

That said, what harm results from letting sqlite reuse a rowid from a deleted row?  Is it that reference attributes will break?  I don't recall if refs enforce integrity or not, but if they don't, then I agree that AUTOINCREMENT is the only safe thing to do.

Reply:
When retrieving a deleted item by storeID (whether from a reference attribute, or a URL, or something else), the expectation is that you will get `ItemNotFound`; that is, storeIDs are meant to be globally unique over the lifetime of the store. I'm not sure if this is explicitly written down in the docs, but there's a lot of software relying on it now.

As an aside: references can't enforce integrity because 1) we don't actually have foreign key constraint checking turned on at all, and 2) references in Axiom are polymorphic, so we don't know which table to have a foreign key constraint against.

We don't particularly care about the rowid _itself_, except insofar as we are using the autoincrement behaviour to allocate/generate the storeID originally by inserting into `axiom_objects`, which is a relatively efficient way to allocate them. In particular, the oid/storeID/rowid is always set explicitly for the item rows themselves, which is why I haven't declared the column AUTOINCREMENT anywhere else besides `axiom_{objects,types}`.

Reply:
Thank you for the explanation, I appreciate it :)

Based on that, I agree that AUTOINCREMENT is the right solution.

Reply:
So regarding the column name; the queries should still use `oid` everywhere to maintain compatibility with databases not upgraded yet. Given that, is it less confusing to actually call the column `storeID` or `oid` in the schema? I think I should probably document this a bit better anyway to avoid land mines for future maintainers.

Reply:
Good point.  While I wish there could be one, and only one, name, it seems like we have to maintain both oid (internal) and storeID (external).  My original thought was that making the column name "storeID" would allow for the eventual removal or paring-down of the storeID property in the Axiom codebase, but it seem like it should remain.

Reply:
Regarding documentation: I think it would be helpful to explicitly note that "oid" is what we use internally, or privately, within Axiom, and "storeID" is what applications should use.